File: swift/Packages/Overstory/Sources/Overstory/DeviceKeys.swift
import CryptoKit
import Foundation

/// A key device's public key as `devices.yaml` spells it (accounts §3): an
/// algorithm tag and the raw public key in unpadded base64url. `ed25519` keys
/// are 32 bytes; `p256` keys are 33-byte compressed SEC1 points, which sign
/// with ECDSA over SHA-256 as the 64-byte concatenation of `r` and `s`.
public struct ProtocolDeviceKey: Sendable, Equatable {
    public enum Algorithm: String, Sendable {
        case ed25519
        case p256
    }

    public var algorithm: Algorithm
    public var publicKey: Data

    private static let spkiPrefixes: [Algorithm: [UInt8]] = [
        .ed25519: [0x30, 0x2a, 0x30, 0x05, 0x06, 0x03, 0x2b, 0x65, 0x70, 0x03, 0x21, 0x00],
        .p256: [
            0x30, 0x39, 0x30, 0x13, 0x06, 0x07, 0x2a, 0x86, 0x48, 0xce, 0x3d, 0x02, 0x01, 0x06, 0x08,
            0x2a, 0x86, 0x48, 0xce, 0x3d, 0x03, 0x01, 0x07, 0x03, 0x22, 0x00,
        ],
    ]

    /// Parse a `key` value; throws unless it is canonical.
    public init(_ value: String) throws {
        let parts = value.split(separator: ":", maxSplits: 1, omittingEmptySubsequences: false)
        guard parts.count == 2,
              let algorithm = Algorithm(rawValue: String(parts[0])),
              let bytes = decodeBase64URL(String(parts[1])),
              bytes.count == (algorithm == .ed25519 ? 32 : 33),
              algorithm == .ed25519 || bytes.first == 0x02 || bytes.first == 0x03 else {
            throw ProtocolValidationError.invalidValue("Malformed device key: \(value)")
        }
        self.algorithm = algorithm
        self.publicKey = bytes
    }

    public init(ed25519 key: Curve25519.Signing.PublicKey) {
        algorithm = .ed25519
        publicKey = key.rawRepresentation
    }

    public init(p256 key: P256.Signing.PublicKey) {
        algorithm = .p256
        publicKey = key.compressedRepresentation
    }

    /// The `devices.yaml` spelling.
    public var value: String { "\(algorithm.rawValue):\(encodeBase64URL(publicKey))" }

    /// The DER SubjectPublicKeyInfo, as canopyd loads it.
    public var spki: Data { Data(Self.spkiPrefixes[algorithm]!) + publicKey }

    /// Whether `signature` (64 bytes, unpadded base64url) signs `message`.
    public func verifies(_ signature: String, over message: Data) -> Bool {
        guard let bytes = decodeBase64URL(signature), bytes.count == 64 else { return false }
        switch algorithm {
        case .ed25519:
            guard let key = try? Curve25519.Signing.PublicKey(rawRepresentation: publicKey) else { return false }
            return key.isValidSignature(bytes, for: message)
        case .p256:
            guard let key = try? P256.Signing.PublicKey(compressedRepresentation: publicKey),
                  let parsed = try? P256.Signing.ECDSASignature(rawRepresentation: bytes) else { return false }
            return key.isValidSignature(parsed, for: message)
        }
    }
}

/// What a key device signs to open a session at `origin` (accounts §5.1).
public struct ProtocolDeviceSessionChallenge: Codable, Sendable, Equatable {
    public var version: Int
    public var purpose: String
    public var id: String
    public var origin: String
    public var profileTree: String
    public var device: String
    public var nonce: String
    public var issuedAt: Int
    public var expiresAt: Int

    public func validated() throws -> Self {
        guard version == 1, purpose == "device-session",
              id.range(of: #"^ax_[a-z2-7]{26}$"#, options: .regularExpression) != nil,
              isCanonicalOrigin(origin),
              profileTree.range(of: #"^tr_[a-z2-7]+$"#, options: .regularExpression) != nil,
              device.range(of: #"^dv_[a-z2-7]+$"#, options: .regularExpression) != nil,
              nonce.range(of: #"^[A-Za-z0-9_-]{43}$"#, options: .regularExpression) != nil,
              issuedAt >= 0, expiresAt > issuedAt else {
            throw ProtocolValidationError.invalidValue("Malformed device session challenge")
        }
        return self
    }
}

/// Exact canonical bytes a key device signs to open a session.
public func deviceSessionChallengeSigningBytes(_ challenge: ProtocolDeviceSessionChallenge) throws -> Data {
    let value = try challenge.validated()
    return CanonicalCBOR.encode(.map([
        ("version", .unsigned(value.version)),
        ("purpose", .text(value.purpose)),
        ("id", .text(value.id)),
        ("origin", .text(value.origin)),
        ("profileTree", .text(value.profileTree)),
        ("device", .text(value.device)),
        ("nonce", .text(value.nonce)),
        ("issuedAt", .unsigned(value.issuedAt)),
        ("expiresAt", .unsigned(value.expiresAt)),
    ]))
}

/// A session a key device opened; `token` is a bearer credential for one host.
public struct ProtocolDeviceSession: Codable, Sendable, Equatable {
    public var token: String
    public var device: String
    public var expiresAt: Int
}

/// The new administrator device a profile-key reset installs.
public struct ProtocolProfileResetDevice: Codable, Sendable, Equatable {
    public var id: String
    public var label: String
    public var key: String

    public init(id: String, label: String, key: String) {
        self.id = id
        self.label = label
        self.key = key
    }

    public func validated() throws -> Self {
        let trimmed = label.trimmingCharacters(in: .whitespacesAndNewlines)
        guard id.range(of: #"^dv_[a-z2-7]{26}$"#, options: .regularExpression) != nil,
              !trimmed.isEmpty, label.utf16.count <= 100,
              (try? ProtocolDeviceKey(key)) != nil else {
            throw ProtocolValidationError.invalidValue("A reset names a new device with a generated DeviceID, a label and a key")
        }
        return self
    }
}

/// What the profile key signs to reset its devices at `origin` (accounts §5.3).
public struct ProtocolProfileResetChallenge: Codable, Sendable, Equatable {
    public var version: Int
    public var purpose: String
    public var id: String
    public var origin: String
    public var profileTree: String
    public var device: ProtocolProfileResetDevice
    public var nonce: String
    public var issuedAt: Int
    public var expiresAt: Int

    public func validated() throws -> Self {
        guard version == 1, purpose == "profile-reset",
              id.range(of: #"^ax_[a-z2-7]{26}$"#, options: .regularExpression) != nil,
              isCanonicalOrigin(origin),
              profileTree.range(of: #"^tr_[a-z2-7]{52}$"#, options: .regularExpression) != nil,
              nonce.range(of: #"^[A-Za-z0-9_-]{43}$"#, options: .regularExpression) != nil,
              issuedAt >= 0, expiresAt > issuedAt else {
            throw ProtocolValidationError.invalidValue("Malformed profile reset challenge")
        }
        _ = try device.validated()
        return self
    }
}

/// Exact canonical bytes the profile key signs to request a reset.
public func profileResetChallengeSigningBytes(_ challenge: ProtocolProfileResetChallenge) throws -> Data {
    let value = try challenge.validated()
    return CanonicalCBOR.encode(.map([
        ("version", .unsigned(value.version)),
        ("purpose", .text(value.purpose)),
        ("id", .text(value.id)),
        ("origin", .text(value.origin)),
        ("profileTree", .text(value.profileTree)),
        ("device", .map([
            ("id", .text(value.device.id)),
            ("label", .text(value.device.label)),
            ("key", .text(value.device.key)),
        ])),
        ("nonce", .text(value.nonce)),
        ("issuedAt", .unsigned(value.issuedAt)),
        ("expiresAt", .unsigned(value.expiresAt)),
    ]))
}

/// A pending reset, as the home host reports it to the profile's devices.
public struct ProtocolPendingProfileReset: Codable, Sendable, Equatable {
    public struct Device: Codable, Sendable, Equatable {
        public var id: String
        public var label: String
    }

    public var profileTree: String
    public var device: Device
    public var requestedAt: Int
    public var effectiveAt: Int
}

private func isCanonicalOrigin(_ value: String) -> Bool {
    guard let url = URL(string: value), let scheme = url.scheme, let host = url.host(), url.path.isEmpty,
          url.query == nil, url.fragment == nil, url.user == nil else { return false }
    let port = url.port.map { ":\($0)" } ?? ""
    return value == "\(scheme.lowercased())://\(host.lowercased())\(port)"
}

func decodeBase64URL(_ value: String) -> Data? {
    guard value.range(of: #"^[A-Za-z0-9_-]*$"#, options: .regularExpression) != nil else { return nil }
    var base64 = value.replacingOccurrences(of: "-", with: "+").replacingOccurrences(of: "_", with: "/")
    base64 += String(repeating: "=", count: (4 - base64.count % 4) % 4)
    guard let data = Data(base64Encoded: base64), encodeBase64URL(data) == value else { return nil }
    return data
}

func encodeBase64URL(_ data: Data) -> String {
    data.base64EncodedString()
        .replacingOccurrences(of: "+", with: "-")
        .replacingOccurrences(of: "/", with: "_")
        .replacingOccurrences(of: "=", with: "")
}

File: packages/protocol/src/model/device-keys.ts
import { encodeCanonicalCBOR } from "./cbor.ts";
import { isGeneratedArborID, isPersonProfileTreeID } from "./identity.ts";

/**
 * A key device's public key as `devices.yaml` spells it: an algorithm tag and
 * the raw public key in unpadded base64url. `ed25519` keys are 32 bytes;
 * `p256` keys are 33-byte compressed SEC1 points, signing with ECDSA over
 * SHA-256 as the 64-byte concatenation of `r` and `s`.
 */
export type DeviceKeyAlgorithm = "ed25519" | "p256";

export interface DeviceKey {
  algorithm: DeviceKeyAlgorithm;
  publicKey: Uint8Array;
}

const KEY_LENGTHS: Record<DeviceKeyAlgorithm, number> = { ed25519: 32, p256: 33 };
/** DER SubjectPublicKeyInfo prefixes; the raw key follows. */
const SPKI_PREFIXES: Record<DeviceKeyAlgorithm, string> = {
  ed25519: "302a300506032b6570032100",
  p256: "3039301306072a8648ce3d020106082a8648ce3d030107032200",
};

function base64urlBytes(value: string): Uint8Array | null {
  if (!/^[A-Za-z0-9_-]*$/.test(value)) return null;
  const padded = value.replaceAll("-", "+").replaceAll("_", "/") + "=".repeat((4 - value.length % 4) % 4);
  let binary: string;
  try { binary = atob(padded); } catch { return null; }
  const bytes = Uint8Array.from(binary, (character) => character.charCodeAt(0));
  return base64url(bytes) === value ? bytes : null;
}

function base64url(bytes: Uint8Array): string {
  let binary = "";
  for (const byte of bytes) binary += String.fromCharCode(byte);
  return btoa(binary).replaceAll("+", "-").replaceAll("/", "_").replace(/=+$/, "");
}

/** Parse a `key` value; throws unless it is canonical. */
export function parseDeviceKey(value: string): DeviceKey {
  const match = /^(ed25519|p256):(.+)$/.exec(value);
  const algorithm = match?.[1] as DeviceKeyAlgorithm | undefined;
  const publicKey = match ? base64urlBytes(match[2]!) : null;
  if (!algorithm || !publicKey || publicKey.byteLength !== KEY_LENGTHS[algorithm]) throw new Error(`Malformed device key: ${value}`);
  if (algorithm === "p256" && publicKey[0] !== 2 && publicKey[0] !== 3) throw new Error(`Malformed device key: ${value}`);
  return { algorithm, publicKey };
}

export function isDeviceKey(value: unknown): value is string {
  if (typeof value !== "string") return false;
  try { parseDeviceKey(value); return true; } catch { return false; }
}

export function formatDeviceKey(key: DeviceKey): string {
  return `${key.algorithm}:${base64url(key.publicKey)}`;
}

/** The key's DER SubjectPublicKeyInfo, which verifiers load. A `p256` key
 * stays compressed; OpenSSL and CryptoKit both accept that form. */
export function deviceKeySPKI(value: string): Uint8Array {
  const key = parseDeviceKey(value);
  const prefix = SPKI_PREFIXES[key.algorithm];
  const bytes = new Uint8Array(prefix.length / 2 + key.publicKey.byteLength);
  for (let index = 0; index < prefix.length / 2; index++) bytes[index] = Number.parseInt(prefix.slice(index * 2, index * 2 + 2), 16);
  bytes.set(key.publicKey, prefix.length / 2);
  return bytes;
}

/** Whether a signature has the shape its key's algorithm produces: 64 bytes,
 * unpadded base64url, for both algorithms. */
export function deviceSignatureBytes(signature: string): Uint8Array {
  const bytes = base64urlBytes(signature);
  if (!bytes || bytes.byteLength !== 64) throw new Error("Malformed device signature");
  return bytes;
}

function nonceIsValid(value: unknown): value is string {
  return typeof value === "string" && /^[A-Za-z0-9_-]{43}$/.test(value);
}

function timesAreValid(issuedAt: unknown, expiresAt: unknown): boolean {
  return typeof issuedAt === "number" && Number.isSafeInteger(issuedAt)
    && typeof expiresAt === "number" && Number.isSafeInteger(expiresAt) && expiresAt > issuedAt;
}

function originIsCanonical(value: unknown): value is string {
  if (typeof value !== "string") return false;
  try { return new URL(value).origin === value; } catch { return false; }
}

function exactKeys(value: object, keys: string[]): boolean {
  const present = Object.keys(value);
  return present.length === keys.length && keys.every((key) => present.includes(key));
}

/** What a key device signs to open a session at `origin` (accounts §5.1). */
export interface DeviceSessionChallenge {
  version: 1;
  purpose: "device-session";
  id: string;
  origin: string;
  profileTree: string;
  device: string;
  nonce: string;
  issuedAt: number;
  expiresAt: number;
}

export function validateDeviceSessionChallenge(value: unknown): DeviceSessionChallenge {
  const fail = () => new Error("Malformed device session challenge");
  if (!value || typeof value !== "object" || Array.isArray(value)) throw fail();
  const challenge = value as Partial<DeviceSessionChallenge>;
  if (
    !exactKeys(challenge, ["version", "purpose", "id", "origin", "profileTree", "device", "nonce", "issuedAt", "expiresAt"])
    || challenge.version !== 1 || challenge.purpose !== "device-session"
    || typeof challenge.id !== "string" || !isGeneratedArborID(challenge.id, "ax")
    || !originIsCanonical(challenge.origin)
    // Any person profile, including one whose TreeID predates self-certifying IDs.
    || typeof challenge.profileTree !== "string" || !/^tr_[a-z2-7]+$/.test(challenge.profileTree)
    || typeof challenge.device !== "string" || !/^dv_[a-z2-7]+$/.test(challenge.device)
    || !nonceIsValid(challenge.nonce) || !timesAreValid(challenge.issuedAt, challenge.expiresAt)
  ) throw fail();
  return challenge as DeviceSessionChallenge;
}

/** Exact bytes a key device signs to open a session. */
export function deviceSessionChallengeBytes(challenge: DeviceSessionChallenge): Uint8Array {
  return encodeCanonicalCBOR(validateDeviceSessionChallenge(challenge));
}

/** A session a key device opened; `token` is a bearer credential for one host. */
export interface DeviceSession {
  token: string;
  device: string;
  expiresAt: number;
}

/** The new administrator device a profile-key reset installs. */
export interface ProfileResetDevice {
  id: string;
  label: string;
  key: string;
}

/** What the profile key signs to reset its devices at `origin` (accounts §5.3). */
export interface ProfileResetChallenge {
  version: 1;
  purpose: "profile-reset";
  id: string;
  origin: string;
  profileTree: string;
  device: ProfileResetDevice;
  nonce: string;
  issuedAt: number;
  expiresAt: number;
}

export function validateProfileResetDevice(value: unknown): ProfileResetDevice {
  const fail = () => new Error("A reset names a new device with a generated DeviceID, a label and a key");
  if (!value || typeof value !== "object" || Array.isArray(value)) throw fail();
  const device = value as Partial<ProfileResetDevice>;
  if (
    !exactKeys(device, ["id", "label", "key"])
    || typeof device.id !== "string" || !isGeneratedArborID(device.id, "dv")
    || typeof device.label !== "string" || !device.label.trim() || device.label.length > 100
    || !isDeviceKey(device.key)
  ) throw fail();
  return device as ProfileResetDevice;
}

export function validateProfileResetChallenge(value: unknown): ProfileResetChallenge {
  const fail = () => new Error("Malformed profile reset challenge");
  if (!value || typeof value !== "object" || Array.isArray(value)) throw fail();
  const challenge = value as Partial<ProfileResetChallenge>;
  if (
    !exactKeys(challenge, ["version", "purpose", "id", "origin", "profileTree", "device", "nonce", "issuedAt", "expiresAt"])
    || challenge.version !== 1 || challenge.purpose !== "profile-reset"
    || typeof challenge.id !== "string" || !isGeneratedArborID(challenge.id, "ax")
    || !originIsCanonical(challenge.origin)
    || typeof challenge.profileTree !== "string" || !isPersonProfileTreeID(challenge.profileTree)
    || !nonceIsValid(challenge.nonce) || !timesAreValid(challenge.issuedAt, challenge.expiresAt)
  ) throw fail();
  validateProfileResetDevice(challenge.device);
  return challenge as ProfileResetChallenge;
}

/** Exact bytes the profile key signs to request a reset. */
export function profileResetChallengeBytes(challenge: ProfileResetChallenge): Uint8Array {
  return encodeCanonicalCBOR(validateProfileResetChallenge(challenge));
}

/** A pending reset, as the home host reports it to the profile's devices. */
export interface PendingProfileReset {
  profileTree: string;
  device: { id: string; label: string };
  requestedAt: number;
  effectiveAt: number;
}

/** One entry of a home host's published device keys (accounts §5.4). */
export interface PublishedDeviceKey {
  id: string;
  key: string;
  administrator: boolean;
}

export interface PublishedDeviceKeys {
  profileTree: string;
  devices: PublishedDeviceKey[];
}
